ARMmbed/mbed-os
# fix #3863 Add an mbed API that allows the init of the CAN at the bus frequency

## Description
If the bus is using a much higher frequency than the default value chosen in the init function, the device cannot synchronize and is blocked (see #3863  for details). The device needs an IDLE frame on the CAN bus to synchronize, but if the device is not set with the correct frequency, the IDLE frame may never be detected.
for instance: initialize CAN at 100kbps. It needs an idle frame on the CAN bus to synchronize (ie: 110us to have 11 consecutive recessive bits). If the bus is at 500kbs and highly busy due to another device, 110us of consecutive recessive bits does not occur. I have only 58us in my use case.
Adding the frequency parameter to the CAN class (same as done in serial api for baud rates) allows us to initialize and synchronize the CAN  device even if the bus is often busy.

## Status
**READY**

## Steps to test or reproduce
Have a 1st device with high bus occupation with the following code:
```
#include "mbed.h"
DigitalOut led2(LED2);
Serial pc(USBTX,USBRX);

int main()
{
    int mycnt=0;
    pc.baud(115200);
    pc.printf("main()\r\n");
    CAN can1(PA_11, PA_12); // only for NUCLEO_F446RE
    pc.printf("created can\r\n");
    can1.frequency(500000);
    pc.printf("set frequency \r\n");

    unsigned int id = 0x539;
    char i =0xC7;
    CANMessage tx_msg(id, &i, sizeof(i));
    bool sent = false;

    while(1) {
        if (can1.write(tx_msg)) {
            mycnt++;
            printf("Sent %u: %d\n", id, i);
            sent = true;
        }
        wait_ms(10);
        led2 = !led2;
    }
}

```
Have the current device trying to initialize. Without the PR, NUCLEO_F767ZI cannot exit from the init function (because HW cannot syncrhonize). With the PR, it can synchronize if the init function is modifying the frequency (we can see the printf "created can") with the following code:
```
#include "mbed.h"
DigitalOut led2(LED2);
Serial pc(USBTX,USBRX);

int main()
{
    pc.baud(115200);
    pc.printf("main()\r\n");
    CAN can2(PD_0, PD_1); //without the PR
    CAN can2(PD_0, PD_1, 500000); //with the PR
    pc.printf("created can\r\n");
    can2_ptr = &can2;
    //can2.frequency(500000); // without the PR
    pc.printf("set frequency \r\n");
    CANMessage msg;

    while(1) {
        if(can2.read(msg)) {
            pc.printf("<-*******Message received:id=%x,data=",msg.id);
            for(int i=0; i<8; i++) {
                pc.printf("%x,",msg.data[i] );
            }
            pc.printf("\n\r");
            led2 = !led2;
        }
    }
}

```

cc @bperry730 @ohagendorf
